jackson-core/src/main/java/io/micronaut/jackson/core/tree/MicronautTreeCodec.java

```diff
@@ -0,0 +1,134 @@
+            case VALUE_STRING:
+                return JsonNode.createStringNode(p.getText());
+            case VALUE_NUMBER_INT:
+                if (config.useBigIntegerForInts()) {
+                    return JsonNode.createNumberNode(p.getBigIntegerValue());
+                } else {
+                    // technically, we could get an unsupported number value here.
+                    return JsonNode.createNumberNodeImpl(p.getNumberValue());
+                }
+            case VALUE_NUMBER_FLOAT:
+                if (config.useBigDecimalForFloats()) {
+                    return JsonNode.createNumberNode(p.getDecimalValue());
+                } else {
+                    // technically, we could get an unsupported number value here.
+                    return JsonNode.createNumberNodeImpl(p.getNumberValue());
+                }
+            case VALUE_TRUE:
+                return JsonNode.createBooleanNode(true);
+            case VALUE_FALSE:
+                return JsonNode.createBooleanNode(false);
+            case VALUE_NULL:
+                return JsonNode.nullNode();
+            default:
+                throw new UnsupportedOperationException("Unsupported token: " + p.currentToken());
+        }
+    }
+
+    public void writeTree(JsonGenerator generator, JsonNode tree) throws IOException, JsonProcessingException {
+        if (tree.isObject()) {
+            generator.writeStartObject();
+            for (Map.Entry<String, JsonNode> entry : tree.entries()) {
+                generator.writeFieldName(entry.getKey());
+                writeTree(generator, tree);
+            }
+            generator.writeEndObject();
+        } else if (tree.isArray()) {
+            generator.writeStartArray();
+            for (JsonNode value : tree.values()) {
+                writeTree(generator, value);
+            }
+            generator.writeEndArray();
+        } else if (tree.isBoolean()) {
+            generator.writeBoolean(tree.getBooleanValue());
+        } else if (tree.isNull()) {
+            generator.writeNull();
+        } else if (tree.isNumber()) {
+            Number value = tree.getNumberValue();
+            if (value instanceof BigDecimal) {
+                generator.writeNumber((BigDecimal) value);
+            } else if (value instanceof Double) {
+                generator.writeNumber(value.doubleValue());
+            } else if (value instanceof Float) {
+                generator.writeNumber(value.floatValue());
+            } else if (value instanceof Integer) {
+                generator.writeNumber(value.intValue());
+            } else if (value instanceof Byte || value instanceof Short) {
+                generator.writeNumber(value.shortValue());
+            } else if (value instanceof Long) {
+                generator.writeNumber(value.longValue());
+            } else if (value instanceof BigInteger) {
+                generator.writeNumber((BigInteger) value);
+            } else {
```

Comment:
order `instanceof` checks by probability. `Integer` and `Long` are the most commonly used should be at the top